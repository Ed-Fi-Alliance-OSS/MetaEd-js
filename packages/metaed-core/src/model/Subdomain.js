// @flow
import { newDomain } from './Domain';
import type { Domain } from './Domain';
import type { SourceMap } from './SourceMap';
import { NoSourceMap } from './SourceMap';
import type { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import type { DomainItem } from './DomainItem';
import type { ModelBase } from './ModelBase';

export type SubdomainSourceMap = {
  ...$Exact<TopLevelEntitySourceMap>,
  domainItems: Array<SourceMap>,
  entities: Array<SourceMap>,
  parent: ?SourceMap,
  parentMetaEdName: ?SourceMap,
  position: ?SourceMap,
};

/**
 *
 */
export function newSubdomainSourceMap(): SubdomainSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    domainItems: [],
    entities: [],
    parent: NoSourceMap,
    parentMetaEdName: NoSourceMap,
    position: NoSourceMap,
  };
}

export type Subdomain = {
  sourceMap: SubdomainSourceMap,
  ...$Exact<TopLevelEntity>,
  domainItems: Array<DomainItem>,
  entities: Array<TopLevelEntity>,
  parent: Domain,
  parentMetaEdName: string,
  position: number,
};

/**
 *
 */
export function newSubdomain(): Subdomain {
  return {
    ...newTopLevelEntity(),
    type: 'subdomain',
    typeHumanizedName: 'Subdomain',
    domainItems: [],
    entities: [],
    parent: newDomain(),
    parentMetaEdName: '',
    position: 0,
    sourceMap: newSubdomainSourceMap(),
  };
}

/**
 *
 */
export const asSubdomain = (x: ModelBase): Subdomain => ((x: any): Subdomain);
