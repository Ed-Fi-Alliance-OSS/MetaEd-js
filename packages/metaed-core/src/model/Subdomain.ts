import { newDomain } from './Domain';
import { Domain } from './Domain';
import { SourceMap } from './SourceMap';
import { NoSourceMap } from './SourceMap';
import { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import { DomainItem } from './DomainItem';
import { ModelBase } from './ModelBase';

export interface SubdomainSourceMap extends TopLevelEntitySourceMap {
  domainItems: Array<SourceMap>;
  entities: Array<SourceMap>;
  parent: SourceMap;
  parentMetaEdName: SourceMap;
  position: SourceMap;
}

/**
 *
 */
export function newSubdomainSourceMap(): SubdomainSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    domainItems: [],
    entities: [],
    parent: NoSourceMap,
    parentMetaEdName: NoSourceMap,
    position: NoSourceMap,
  };
}

export interface Subdomain extends TopLevelEntity {
  sourceMap: SubdomainSourceMap;
  domainItems: Array<DomainItem>;
  entities: Array<TopLevelEntity>;
  parent: Domain;
  parentMetaEdName: string;
  position: number;
}

/**
 *
 */
export function newSubdomain(): Subdomain {
  return {
    ...newTopLevelEntity(),
    type: 'subdomain',
    typeHumanizedName: 'Subdomain',
    domainItems: [],
    entities: [],
    parent: newDomain(),
    parentMetaEdName: '',
    position: 0,
    sourceMap: newSubdomainSourceMap(),
  };
}

/**
 *
 */
export const asSubdomain = (x: ModelBase): Subdomain => x as Subdomain;
