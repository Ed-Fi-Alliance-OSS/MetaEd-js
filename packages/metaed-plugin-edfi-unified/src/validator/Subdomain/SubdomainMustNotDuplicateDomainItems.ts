import { MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { findDuplicates } from '../ValidatorShared/FindDuplicates';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.subdomain.forEach(subdomain => {
      const names = subdomain.domainItems.map(di => di.metaEdName);
      const duplicates: Array<string> = findDuplicates(names);

      duplicates.forEach(val => {
        const domainItem = subdomain.domainItems.find(d => d.metaEdName === val);
        if (domainItem !== undefined) {
          failures.push({
            validatorName: 'SubdomainMustNotDuplicateDomainItems',
            category: 'error',
            message: `Subdomain Entity Domain Item '${domainItem.metaEdName}' has a duplicate Domain Item.`,
            sourceMap: domainItem.sourceMap.type,
            fileMap: null,
          });
        }
      });
    });
  });

  return failures;
}
