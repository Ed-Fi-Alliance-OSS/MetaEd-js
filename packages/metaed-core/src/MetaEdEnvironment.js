// @flow
import { newEntityRepository } from './model/EntityRepository';
import { newPropertyIndex } from './model/property/PropertyRepository';
import type { EntityRepository } from './model/EntityRepository';
import type { PluginEnvironment } from './plugin/PluginEnvironment';
import type { PropertyIndex } from './model/property/PropertyRepository';

export type SemVer = string;

// provided to validators and enhancers
export type MetaEdEnvironment = {
  // the entity repository
  entity: EntityRepository,

  // the property index, provided for convenience (a little hesitant here, as it should be dependent on entityRepository)
  propertyIndex: PropertyIndex,

  // plugin environment by plugin name
  plugin: Map<string, PluginEnvironment>,

  // the current MetaEd Version
  metaEdVersion: SemVer,

  // the data standard version to target
  dataStandardVersion: SemVer,
};

export const newMetaEdEnvironment: () => MetaEdEnvironment = () => ({
  entity: newEntityRepository(),
  propertyIndex: newPropertyIndex(),
  plugin: new Map(),
  metaEdVersion: '0.0.0',
  dataStandardVersion: '0.0.0',
});
