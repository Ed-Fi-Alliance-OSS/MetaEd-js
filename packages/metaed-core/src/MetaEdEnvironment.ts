import { newPropertyIndex } from './model/property/PropertyRepository';
import { Namespace } from './model/Namespace';
import { PluginEnvironment } from './plugin/PluginEnvironment';
import { PropertyIndex } from './model/property/PropertyRepository';

/**
 *
 */
export type SemVer = string;

/**
 * Provided to validators and enhancers
 */
export type MetaEdEnvironment = {
  // namespaces
  namespace: Map<string, Namespace>;

  // the property index, provided for convenience (a little hesitant here, as it should be dependent on entityRepository)
  propertyIndex: PropertyIndex;

  // plugin environment by plugin shortName
  plugin: Map<string, PluginEnvironment>;

  // the current MetaEd Version
  metaEdVersion: SemVer;

  // the data standard version to target
  dataStandardVersion: SemVer;
};

/**
 *
 */
export const newMetaEdEnvironment: () => MetaEdEnvironment = () => ({
  namespace: new Map(),
  propertyIndex: newPropertyIndex(),
  plugin: new Map(),
  metaEdVersion: '0.0.0',
  dataStandardVersion: '0.0.0',
});
