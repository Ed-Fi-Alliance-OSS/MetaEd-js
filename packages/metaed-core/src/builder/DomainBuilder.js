// @flow

import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { MetaEdGrammarListener } from '../grammar/gen/MetaEdGrammarListener';

import type { Domain, DomainSourceMap } from '../model/Domain';
import type { Subdomain, SubdomainSourceMap } from '../model/Subdomain';
import type { DomainItem, DomainItemSourceMap } from '../model/DomainItem';
import type { MetaEdEnvironment } from '../MetaEdEnvironment';
import type { NamespaceInfo } from '../model/NamespaceInfo';
import { NoNamespaceInfo, newNamespaceInfo } from '../model/NamespaceInfo';
import { newDomainItem, NoDomainItem } from '../model/DomainItem';
import { newDomain, NoDomain } from '../model/Domain';
import { newSubdomain } from '../model/Subdomain';
import { enteringNamespaceName, enteringNamespaceType } from './NamespaceInfoBuilder';
import { extractDocumentation, squareBracketRemoval, isErrorText } from './BuilderUtility';
import type { ValidationFailure } from '../validator/ValidationFailure';
import { sourceMapFrom } from '../model/SourceMap';

export class DomainBuilder extends MetaEdGrammarListener {
  metaEd: MetaEdEnvironment;
  namespaceInfo: NamespaceInfo;
  currentDomain: Domain | Subdomain;
  currentDomainItem: DomainItem;
  validationFailures: Array<ValidationFailure>;

  constructor(metaEd: MetaEdEnvironment, validationFailures: Array<ValidationFailure>) {
    super();
    this.metaEd = metaEd;
    this.namespaceInfo = NoNamespaceInfo;
    this.currentDomain = NoDomain;
    this.currentDomainItem = NoDomainItem;
    this.validationFailures = validationFailures;
  }

  enterNamespace(context: MetaEdGrammar.NamespaceContext) {
    if (this.namespaceInfo !== NoNamespaceInfo) return;
    this.namespaceInfo = newNamespaceInfo();
    this.namespaceInfo.sourceMap.type = sourceMapFrom(context);
  }

  enterNamespaceName(context: MetaEdGrammar.NamespaceNameContext) {
    if (this.namespaceInfo === NoNamespaceInfo) return;
    this.namespaceInfo = enteringNamespaceName(context, this.namespaceInfo);
  }

  enterNamespaceType(context: MetaEdGrammar.NamespaceTypeContext) {
    if (this.namespaceInfo === NoNamespaceInfo) return;
    this.namespaceInfo = enteringNamespaceType(context, this.namespaceInfo);
  }

  // eslint-disable-next-line no-unused-vars
  exitNamespace(context: MetaEdGrammar.NamespaceContext) {
    this.namespaceInfo = NoNamespaceInfo;
  }

  enterDocumentation(context: MetaEdGrammar.DocumentationContext) {
    if (this.currentDomain === NoDomain) return;
    this.currentDomain.documentation = extractDocumentation(context);
    this.currentDomain.sourceMap.documentation = sourceMapFrom(context);
  }

  enterMetaEdId(context: MetaEdGrammar.MetaEdIdContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.METAED_ID() == null || context.METAED_ID().exception != null || isErrorText(context.METAED_ID().getText())) return;

    if (this.currentDomainItem !== NoDomainItem) {
      this.currentDomainItem.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
      this.currentDomainItem.sourceMap.metaEdId = sourceMapFrom(context);
    } else {
      this.currentDomain.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
      this.currentDomain.sourceMap.metaEdId = sourceMapFrom(context);
    }
  }

  enterDomain(context: MetaEdGrammar.DomainContext) {
    if (this.namespaceInfo === NoNamespaceInfo) return;
    this.currentDomain = Object.assign(newDomain(), { namespaceInfo: this.namespaceInfo });
    this.currentDomain.sourceMap.type = sourceMapFrom(context);
    this.currentDomain.sourceMap.namespaceInfo = this.namespaceInfo.sourceMap.type;
  }

  enterSubdomain(context: MetaEdGrammar.SubdomainContext) {
    if (this.namespaceInfo === NoNamespaceInfo) return;
    this.currentDomain = Object.assign(newSubdomain(), { namespaceInfo: this.namespaceInfo });
    this.currentDomain.sourceMap.type = sourceMapFrom(context);
    this.currentDomain.sourceMap.namespaceInfo = this.namespaceInfo.sourceMap.type;
  }

  exitingEntity() {
    if (this.currentDomain === NoDomain) return;
    // $FlowIgnore
    const currentDomainRepository = this.metaEd.entity[this.currentDomain.type];
    if (this.currentDomain.metaEdName) {
      if (currentDomainRepository.has(this.currentDomain.metaEdName)) {
        this.validationFailures.push({
          validatorName: 'DomainBuilder',
          category: 'error',
          message: `${this.currentDomain.typeHumanizedName} named ${this.currentDomain.metaEdName} is a duplicate declaration of that name.`,
          sourceMap: this.currentDomain.sourceMap.type,
          fileMap: null,
        });
        const duplicateEntity: Domain | Subdomain = currentDomainRepository.get(this.currentDomain.metaEdName);
        this.validationFailures.push({
          validatorName: 'DomainBuilder',
          category: 'error',
          message: `${duplicateEntity.typeHumanizedName} named ${duplicateEntity.metaEdName} is a duplicate declaration of that name.`,
          sourceMap: duplicateEntity.sourceMap.type,
          fileMap: null,
        });
      } else {
        currentDomainRepository.set(this.currentDomain.metaEdName, this.currentDomain);
      }
    }

    this.currentDomain = NoDomain;
  }

  // eslint-disable-next-line no-unused-vars
  exitDomain(context: MetaEdGrammar.DomainContext) {
    this.exitingEntity();
  }

  // eslint-disable-next-line no-unused-vars
  exitSubdomain(context: MetaEdGrammar.DomainContext) {
    this.exitingEntity();
  }

  enterDomainName(context: MetaEdGrammar.DomainNameContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.currentDomain.metaEdName = context.ID().getText();
    this.currentDomain.sourceMap.metaEdName = sourceMapFrom(context);
  }

  enterSubdomainName(context: MetaEdGrammar.SubdomainNameContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.currentDomain.metaEdName = context.ID().getText();
    this.currentDomain.sourceMap.metaEdName = sourceMapFrom(context);
  }

  enterParentDomainName(context: MetaEdGrammar.ParentDomainNameContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    ((this.currentDomain: any): Subdomain).parentMetaEdName = context.ID().getText();
    ((this.currentDomain.sourceMap: any): SubdomainSourceMap).parent = sourceMapFrom(context);
    ((this.currentDomain.sourceMap: any): SubdomainSourceMap).parentMetaEdName = sourceMapFrom(context);
  }

  enterSubdomainPosition(context: MetaEdGrammar.SubdomainPositionContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.UNSIGNED_INT() == null || context.UNSIGNED_INT().exception || isErrorText(context.UNSIGNED_INT().getText())) return;
    ((this.currentDomain: any): Subdomain).position = Number(context.UNSIGNED_INT().getText());
    ((this.currentDomain.sourceMap: any): SubdomainSourceMap).position = sourceMapFrom(context);
  }

  enterDomainItem(context: MetaEdGrammar.DomainItemContext) {
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.currentDomainItem = Object.assign(newDomainItem(), { metaEdName: context.ID().getText() });
    ((this.currentDomainItem.sourceMap: any): DomainItemSourceMap).metaEdName = sourceMapFrom(context);
    ((this.currentDomainItem.sourceMap: any): DomainItemSourceMap).referencedType = sourceMapFrom(context);

    // mutually exclusive in language
    if (context.ASSOCIATION_KEYWORD()) this.currentDomainItem.referencedType = 'association';
    if (context.COMMON_KEYWORD()) this.currentDomainItem.referencedType = 'common';
    if (context.DOMAIN_ENTITY_KEYWORD()) this.currentDomainItem.referencedType = 'domainEntity';
    if (context.DESCRIPTOR_KEYWORD()) this.currentDomainItem.referencedType = 'descriptor';
    if (context.INLINE_COMMON_KEYWORD()) this.currentDomainItem.referencedType = 'inlineCommon';
  }

  exitDomainItem(context: MetaEdGrammar.DomainItemContext) {
    if (this.currentDomain === NoDomain || this.currentDomainItem === NoDomainItem) return;
    ((this.currentDomain.sourceMap: any): DomainSourceMap).domainItems.push(sourceMapFrom(context));
    this.currentDomain.domainItems.push(this.currentDomainItem);
    this.currentDomainItem = NoDomainItem;
  }

  enterFooterDocumentation(context: MetaEdGrammar.FooterDocumentationContext) {
    if (this.currentDomain === NoDomain) return;
    ((this.currentDomain: any): Domain).footerDocumentation = extractDocumentation(context);
    ((this.currentDomain.sourceMap: any): DomainSourceMap).footerDocumentation = sourceMapFrom(context);
  }
}
