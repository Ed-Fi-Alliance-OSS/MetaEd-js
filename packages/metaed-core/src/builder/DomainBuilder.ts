import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { MetaEdGrammarListener } from '../grammar/gen/MetaEdGrammarListener';

import { Domain, DomainSourceMap } from '../model/Domain';
import { Subdomain, SubdomainSourceMap } from '../model/Subdomain';
import { DomainItem } from '../model/DomainItem';
import { MetaEdEnvironment } from '../MetaEdEnvironment';
import { Namespace } from '../model/Namespace';
import { NoNamespace } from '../model/Namespace';
import { newDomainItem, NoDomainItem } from '../model/DomainItem';
import { newDomain, NoDomain } from '../model/Domain';
import { newSubdomain } from '../model/Subdomain';
import { namespaceNameFrom } from './NamespaceBuilder';
import { extractDocumentation, extractDeprecationReason, squareBracketRemoval, isErrorText } from './BuilderUtility';
import { ValidationFailure } from '../validator/ValidationFailure';
import { sourceMapFrom } from '../model/SourceMap';
import { TopLevelEntity } from '../model/TopLevelEntity';

/**
 * An ANTLR4 listener that creates Domain (not DomainEntity!) entities.
 */
export class DomainBuilder extends MetaEdGrammarListener {
  metaEd: MetaEdEnvironment;

  currentNamespace: Namespace;

  currentDomain: Domain | Subdomain;

  currentDomainItem: DomainItem;

  validationFailures: ValidationFailure[];

  constructor(metaEd: MetaEdEnvironment, validationFailures: ValidationFailure[]) {
    super();
    this.metaEd = metaEd;
    this.currentNamespace = NoNamespace;
    this.currentDomain = NoDomain;
    this.currentDomainItem = NoDomainItem;
    this.validationFailures = validationFailures;
  }

  enterNamespaceName(context: MetaEdGrammar.NamespaceNameContext) {
    const namespace: Namespace | undefined = this.metaEd.namespace.get(namespaceNameFrom(context));
    this.currentNamespace = namespace == null ? NoNamespace : namespace;
  }

  enterDeprecated(context: MetaEdGrammar.DeprecatedContext) {
    if (this.currentDomain === NoDomain) return;

    if (!context.exception) {
      this.currentDomain.isDeprecated = true;
      this.currentDomain.deprecationReason = extractDeprecationReason(context);
      this.currentDomain.sourceMap.isDeprecated = sourceMapFrom(context);
      this.currentDomain.sourceMap.deprecationReason = sourceMapFrom(context);
    }
  }

  enterDocumentation(context: MetaEdGrammar.DocumentationContext) {
    if (this.currentDomain === NoDomain) return;
    this.currentDomain.documentation = extractDocumentation(context);
    this.currentDomain.sourceMap.documentation = sourceMapFrom(context);
  }

  enterMetaEdId(context: MetaEdGrammar.MetaEdIdContext) {
    if (this.currentDomain === NoDomain) return;
    if (
      context.exception ||
      context.METAED_ID() == null ||
      context.METAED_ID().exception != null ||
      isErrorText(context.METAED_ID().getText())
    )
      return;

    if (this.currentDomainItem !== NoDomainItem) {
      this.currentDomainItem.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
      this.currentDomainItem.sourceMap.metaEdId = sourceMapFrom(context);
    } else {
      this.currentDomain.metaEdId = squareBracketRemoval(context.METAED_ID().getText());
      this.currentDomain.sourceMap.metaEdId = sourceMapFrom(context);
    }
  }

  enterDomain(context: MetaEdGrammar.DomainContext) {
    this.currentDomain = { ...newDomain(), namespace: this.currentNamespace };
    this.currentDomain.sourceMap.type = sourceMapFrom(context);
  }

  enterSubdomain(context: MetaEdGrammar.SubdomainContext) {
    this.currentDomain = { ...newSubdomain(), namespace: this.currentNamespace };
    this.currentDomain.sourceMap.type = sourceMapFrom(context);
  }

  exitingEntity() {
    if (this.currentDomain === NoDomain) return;

    const currentDomainRepository: Map<string, TopLevelEntity> = this.currentNamespace.entity[this.currentDomain.type];
    if (this.currentDomain.metaEdName) {
      if (currentDomainRepository.has(this.currentDomain.metaEdName)) {
        this.validationFailures.push({
          validatorName: 'DomainBuilder',
          category: 'error',
          message: `${this.currentDomain.typeHumanizedName} named ${this.currentDomain.metaEdName} is a duplicate declaration of that name.`,
          sourceMap: this.currentDomain.sourceMap.metaEdName,
          fileMap: null,
        });
        const duplicateEntity: any = currentDomainRepository.get(this.currentDomain.metaEdName);
        this.validationFailures.push({
          validatorName: 'DomainBuilder',
          category: 'error',
          message: `${duplicateEntity.typeHumanizedName} named ${duplicateEntity.metaEdName} is a duplicate declaration of that name.`,
          sourceMap: duplicateEntity.sourceMap.metaEdName,
          fileMap: null,
        });
      } else {
        currentDomainRepository.set(this.currentDomain.metaEdName, this.currentDomain);
      }
    }

    this.currentDomain = NoDomain;
  }

  // @ts-ignore
  exitDomain(context: MetaEdGrammar.DomainContext) {
    this.exitingEntity();
  }

  // @ts-ignore
  exitSubdomain(context: MetaEdGrammar.DomainContext) {
    this.exitingEntity();
  }

  enterDomainName(context: MetaEdGrammar.DomainNameContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.currentDomain.metaEdName = context.ID().getText();
    this.currentDomain.sourceMap.metaEdName = sourceMapFrom(context);
  }

  enterSubdomainName(context: MetaEdGrammar.SubdomainNameContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.currentDomain.metaEdName = context.ID().getText();
    this.currentDomain.sourceMap.metaEdName = sourceMapFrom(context);
  }

  enterParentDomainName(context: MetaEdGrammar.ParentDomainNameContext) {
    if (this.currentDomain === NoDomain) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    (this.currentDomain as Subdomain).parentMetaEdName = context.ID().getText();
    (this.currentDomain.sourceMap as SubdomainSourceMap).parent = sourceMapFrom(context);
    (this.currentDomain.sourceMap as SubdomainSourceMap).parentMetaEdName = sourceMapFrom(context);
  }

  enterSubdomainPosition(context: MetaEdGrammar.SubdomainPositionContext) {
    if (this.currentDomain === NoDomain) return;
    if (
      context.exception ||
      context.UNSIGNED_INT() == null ||
      context.UNSIGNED_INT().exception ||
      isErrorText(context.UNSIGNED_INT().getText())
    )
      return;
    (this.currentDomain as Subdomain).position = Number(context.UNSIGNED_INT().getText());
    (this.currentDomain.sourceMap as SubdomainSourceMap).position = sourceMapFrom(context);
  }

  enterDomainItem(context: MetaEdGrammar.DomainItemContext) {
    if (this.currentDomain === NoDomain) return;

    if (context.exception || context.localDomainItemName() == null) return;
    const localDomainItemNameContext = context.localDomainItemName();
    if (
      localDomainItemNameContext.exception ||
      localDomainItemNameContext.ID() == null ||
      localDomainItemNameContext.ID().exception ||
      isErrorText(localDomainItemNameContext.ID().getText())
    )
      return;

    this.currentDomainItem = {
      ...newDomainItem(),
      metaEdName: localDomainItemNameContext.ID().getText(),
      namespace: this.currentDomain.namespace,
    };

    Object.assign(this.currentDomainItem.sourceMap, {
      type: sourceMapFrom(localDomainItemNameContext),
      metaEdName: sourceMapFrom(localDomainItemNameContext),
      referencedType: sourceMapFrom(localDomainItemNameContext),
    });

    // mutually exclusive in language
    if (context.ASSOCIATION_KEYWORD()) this.currentDomainItem.referencedType = 'association';
    if (context.COMMON_KEYWORD()) this.currentDomainItem.referencedType = 'common';
    if (context.DOMAIN_ENTITY_KEYWORD()) this.currentDomainItem.referencedType = 'domainEntity';
    if (context.DESCRIPTOR_KEYWORD()) this.currentDomainItem.referencedType = 'descriptor';
    if (context.INLINE_COMMON_KEYWORD()) this.currentDomainItem.referencedType = 'inlineCommon';

    const baseNamespaceContext = context.baseNamespace();
    if (
      baseNamespaceContext == null ||
      baseNamespaceContext.exception ||
      baseNamespaceContext.ID() == null ||
      baseNamespaceContext.ID().exception ||
      isErrorText(baseNamespaceContext.ID().getText())
    ) {
      this.currentDomainItem.referencedNamespaceName = this.currentNamespace.namespaceName;
      this.currentDomainItem.sourceMap.referencedNamespaceName = this.currentDomainItem.sourceMap.metaEdName;
    } else {
      this.currentDomainItem.referencedNamespaceName = baseNamespaceContext.ID().getText();
      this.currentDomainItem.sourceMap.referencedNamespaceName = sourceMapFrom(baseNamespaceContext);
    }
  }

  exitDomainItem(context: MetaEdGrammar.DomainItemContext) {
    if (this.currentDomain === NoDomain || this.currentDomainItem === NoDomainItem) return;
    (this.currentDomain.sourceMap as DomainSourceMap).domainItems.push(sourceMapFrom(context));
    this.currentDomain.domainItems.push(this.currentDomainItem);
    this.currentDomainItem = NoDomainItem;
  }

  enterFooterDocumentation(context: MetaEdGrammar.FooterDocumentationContext) {
    if (this.currentDomain === NoDomain) return;
    (this.currentDomain as Domain).footerDocumentation = extractDocumentation(context);
    (this.currentDomain.sourceMap as DomainSourceMap).footerDocumentation = sourceMapFrom(context);
  }
}
