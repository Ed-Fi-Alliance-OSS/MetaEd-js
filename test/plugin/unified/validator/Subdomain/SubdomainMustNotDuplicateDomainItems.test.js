// @flow
import DomainBuilder from '../../../../../src/core/builder/DomainBuilder';
import MetaEdTextBuilder from '../../../../core/MetaEdTextBuilder';
import { metaEdEnvironmentFactory } from '../../../../../src/core/MetaEdEnvironment';
import type { MetaEdEnvironment } from '../../../../../src/core/MetaEdEnvironment';
import { validate } from '../../../../../src/plugin/unified/validator/Subdomain/SubdomainMustNotDuplicateDomainItems';
import type { ValidationFailure } from '../../../../../src/core/validator/ValidationFailure';

describe('when validating subdomain entity domain item does not duplicate domain items', () => {
  const metaEd: MetaEdEnvironment = metaEdEnvironmentFactory();
  const domainName: string = 'DomainName';
  const subdomainName: string = 'SubdomainName';

  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('namespace', 'ProjectExtension')
      .withStartSubdomain(subdomainName, domainName, '1')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('DomainItem1')
      .withDomainEntityDomainItem('DomainItem2')
      .withEndSubdomain()
      .withEndNamespace()
      .sendToListener(new DomainBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build one subdomain entity', () => {
    expect(metaEd.entity.subdomain.size).toBe(1);
  });

  it('should build two domain entity domain items', () => {
    const subdomain = metaEd.entity.subdomain.get(subdomainName);
    expect(subdomain).toBeDefined();
    if (subdomain !== undefined) {
      expect(subdomain.domainItems.length).toBe(2);
    }
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating subdomain entity domain item has duplicate domain items', () => {
  const metaEd: MetaEdEnvironment = metaEdEnvironmentFactory();
  const domainName: string = 'DomainName';
  const subdomainName: string = 'SubdomainName';
  const domainEntityName: string = 'DomainEntityName';

  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('namespace', 'ProjectExtension')
      .withStartSubdomain(subdomainName, domainName, '1')
      .withDocumentation('doc')
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName)
      .withEndSubdomain()
      .withEndNamespace()
      .sendToListener(new DomainBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build one subdomain entity', () => {
    expect(metaEd.entity.subdomain.size).toBe(1);
  });

  it('should have one validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('SubdomainMustNotDuplicateDomainItems');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when subdomain entity domain item has duplicate should have validation failure -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when subdomain entity domain item has duplicate should have validation failure -> sourceMap');
  });
});

describe('when validating subdomain entity domain item has multiple duplicate domain items', () => {
  const metaEd: MetaEdEnvironment = metaEdEnvironmentFactory();
  const domainName: string = 'DomainName';
  const subdomainName: string = 'SubdomainName';
  const domainEntityName: string = 'DomainEntityName';
  const domainEntityName2: string = 'DomainEntityName2';

  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('namespace', 'ProjectExtension')
      .withStartSubdomain(subdomainName, domainName, '1')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('NotDuplicate')
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName)
      .withDomainEntityDomainItem(domainEntityName2)
      .withDomainEntityDomainItem(domainEntityName2)
      .withEndSubdomain()
      .withEndNamespace()
      .sendToListener(new DomainBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build one subdomain entity', () => {
    expect(metaEd.entity.subdomain.size).toBe(1);
  });

  it('should have multiple validation failures', () => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('SubdomainMustNotDuplicateDomainItems');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when subdomain entity domain item has duplicate should have validation failure -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when subdomain entity domain item has duplicate should have validation failure -> sourceMap');
    expect(failures[1].validatorName).toBe('SubdomainMustNotDuplicateDomainItems');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot('when subdomain entity domain item has duplicate should have validation failure -> message');
    expect(failures[1].sourceMap).toMatchSnapshot('when subdomain entity domain item has duplicate should have validation failure -> sourceMap');
  });
});

