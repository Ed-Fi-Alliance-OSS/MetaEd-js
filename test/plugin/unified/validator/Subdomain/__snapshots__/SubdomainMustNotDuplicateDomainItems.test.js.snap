// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when subdomain entity domain item has duplicate should have validation failure -> message 1`] = `"Subdomain Entity Domain Item 'DomainEntityName' has a duplicate Domain Item."`;

exports[`when subdomain entity domain item has duplicate should have validation failure -> message 2`] = `"Subdomain Entity Domain Item 'DomainEntityName' has a duplicate Domain Item."`;

exports[`when subdomain entity domain item has duplicate should have validation failure -> message 3`] = `"Subdomain Entity Domain Item 'DomainEntityName2' has a duplicate Domain Item."`;

exports[`when subdomain entity domain item has duplicate should have validation failure -> sourceMap 1`] = `undefined`;

exports[`when subdomain entity domain item has duplicate should have validation failure -> sourceMap 2`] = `undefined`;

exports[`when subdomain entity domain item has duplicate should have validation failure -> sourceMap 3`] = `undefined`;
